File: Data Structures & Algorithms/eating-bananas/submission-1.js
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */
    minEatingSpeed(piles, h) {
        let left = 0;
        let right = Math.max(...piles);
        let res = right;

        while (left <= right) {
            let mid = Math.floor((left+right)/2);
            let hours = 0;

            for (let pile of piles) {
                hours += Math.ceil(pile/mid);
            }

            if (hours <= h) {
                res = Math.min(res, mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/time-based-key-value-store/submission-1.py
class TimeMap:

    def __init__(self):
        self.key_store = {} # pair: key -> list of [value, timestamp]

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.key_store:
            self.key_store[key] = []
        self.key_store[key].append([value, timestamp])

    def get(self, key: str, timestamp: int) -> str:
        res = ""
        values = self.key_store.get(key, [])
        left, right = 0, len(values)-1

        while left <= right:
            mid = (left+right) // 2
            
            if values[mid][1] <= timestamp:
                res = values[mid][0]
                left = mid + 1
            else:
                right = mid - 1

        return res